README.md
# psp-backend

NodeJS backend application using the Express framework.

## Requirements

Make sur you have both [npm](https://www.npmjs.com/) and [Node.js](https://nodejs.org/en/) installed.

## Installation

**psp-backend** is very easy to install. Simply follow the following steps:

- Clone this repository in the command line
- Type `npm install` and press enter to install all the required dependencies
- Run the backend by typing `npm start`

## Supported API methods

- user:
  - register (POST /api/register) -> register (login and password in body)
  - authenticate (POST /api/authenticate) -> login and password in body, get a cookie when authentified
  - logout (POST /api/logout) -> logout user, delete sessions and request frontend to delete cookie
- device/tracker:
  - register (POST /api/devices/{eui}) -> register a new device for the authenticated user (device eui in GET)
  - get devices (GET /api/devices) -> get all devices registered by authenticated user
  - get device (GET /api/devices/{eui}) -> get device with {eui} from authenticated user
  - delete (DELETE /api/devices/{eui}) -> unregister device with {eui} from authenticated user

## Response types

The API **always** sends back a JSON object to the frontend client and sets the appropriate HTTP status code so that the client can more easily analyze how the request went


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2021-12-06

### Added

- Created backend project
- API call to campusIot to check if a device exists

## [1.0.1] - 2021-12-23

### Added

- Implemented **/api/users/register** route in API:

  - Declaration of **RegEx** validators in [validator.js](./validator.js) to validate user inputs
  - Checks for already existing user with same username/email
  - Usage of scure **bcrypt** hashing function to hash the user's password
  - Update of the user JSON file with the new user

- Started implementing **/api/users/authenticate** route in API:

  - Implemented credentials check using username/email comparisons and **bcrypt** to compare submitted password
  - **_TODO: Set a session Cookie server-side to keep track of the authentified user and send them the cookie_**

- Added default route in API to redirect to 404

## [1.0.2] - 2021-12-24

### Added

- Implemented cookies using the **express-session** package:
  - Added a specific server-side cookie to keep track of authenticated users using the given session ID
- Added error logging for **bcrypt** related function calls (salt generation + hash)
- Implemented **/api/devices/register** route in API:
  - Added new validator **isEIDValid** in [validator.js](./validator.js) to validate sent EIDs
  - Added the possibility for an authenticated user to register a device to their account
  - **_TODO:_ Make an API call to CampusIoT to check that the device is a legitimate one**

## [1.0.3] - 2022-01-06

### Added

- Added new shortucut command `npm run start` in [package.json](./package.json) to start the backend
- Added logging function in [novlog.js](./novlog.js)
- Added new documentation file [api_test.md](./api_test.md) to easily test out the API
- Implemented new functions in [index.js](./index.js):
  - The backend app now uses the [Axios library](https://axios-http.com/) instead of native HTTP functions
  - Implementend new method **getJWT** to automatically obtain the JWT on startup
  - Changed route **POST api/devices/register** to **POST api/devices/:eid** (the device ID is now passed through GET params instead of body):
    - The backend now checks with the remote LoRaWan endpoint to see whether the device is legitimate
  - Implemented route **DELETE api/devices/:eid** to allow an authenticated user to remove a registered device
  - Implemented route **GET api/devices** to allow an authenticated user to get the eids of all their devices
  - Normalized API responses (JSON objects with proper HTTP status set)
  - Refactored blocks of code intended to write to a file into the function **writeJSONToFile**

### Changed

- Updated [README](./README.md)
- Refactored configuration format in [config.json](./config.json)

### Fixed

- Fixed regular expression in function **isEIDValid** in [validator.js](./validator.js)
- Fixed error message in function **validateUserRegisterForm** in [validator.js](./validator.js)

## [1.1.0] - 2022-01-07

### Added

- Added new command line in [api_test.md](./api_test.md) for newly added route **GET /api/devices/:eui**
- Added new configuration options in [config.json](./config.json):
  - **deviceMaxPos**: maximum number of cached device positions in server file [devices.json](./data/devices.json)
  - **delay**: API call cooldown to LoRaWan endpoint (the greater the better). Used to limit API workload
- Improved logging component [novlog.js](./novlog.js):
  - Added additional log information (date, time, log type)
  - On top of log messages appearing in the console, all transactions are logged to a specific daily log file in [logs](./logs) folder
  - Renamed and refactored log functions
  - Added new log functions for different log types
- Added log messages in [index.js](./index.js)
- Started implementing route **GET /api/devices/:eui** in [index.js](./index.js):
  - Currently checks timestamps of stored device positions and compare them to current time to determine whether or not the information has become obsolete (delay can be modified in [config.json](./config.json))
  - **_TODO: Make correct API call to LoRaWan endpoint to get actual GPS data_**

### Changed

- Moved credentials from [config.json](./config.json) to [credentials.json](./credentials.json) (untracked in [.gitignore](./.gitignore))
- Fixed typo (eid -> eui):
  - Updated occurences in [README](./README.md) and [index.js](./index.js)
  - Renamed function **isEIDValid** to **isEUIValid** in [validator.js](./validator.js)
- Refactored authentication checks in [index.js](./index.js) into a seperate method **verifyAuthentication**

## [1.2.0] - 2022-01-22

### Added

- Added new validator **validateUserAuthenticateForm** in [validator.js](./validator.js) to validate authentication forms
- Added **webapp** configuration in [config.json](./config.json)
- Added **CORS** between **server** and **webapp**
- Added authentication form checks for route **/api/authenticate** in [index.js](./index.js)
- Added new route **POST /api/logout** in [index.js](./index.js)

### Changed

- Changed existing routes in [index.js](./index.js):
  - **POST /api/users/register** was changed to **POST /api/register**
  - **POST /api/users/authenticate** was changed to **POST /api/authenticate**
  - Updated tests in [api_test.md](./api_test.md) accordingly
  - Temporarily hardcoded fake GPS data in **GET /api/devices/:eui([0-9a-fA-F]{16})** for **webapp** testing
- Updated [README](./README.md) to list the newly implemented **logout** route

## [1.3.0] - 2022-01-24

### Added

- Added new service [peter](./peter.js) to generate fake pet data

### Changed

- Added config for the **peter** service in [config.json](./config.json)
- Replaced hardcoded data in [index.js](./index.js) by call to **peter** service


peter.js
const cfg = require("./config.json");
const states = cfg.peter.states;
const WALK = states.find((st) => st.name === "WALK");

let pet = {
  state: WALK.name,
  speed: WALK.speed,
  minProb: WALK.minProb,
  streak: WALK.streak,
};

// define a start pos (Polygone building in Grenoble)
let pos = {
  latitude: 45.20415,
  longitude: 5.6933013,
};

function getPeterPos() {
  pos = calculateNextPos(pos);
  return {
    latitude: pos.latitude,
    longitude: pos.longitude,
    timestamp: new Date(),
  };
}

function calculateNextPos(pos) {
  // If streak reaches 0, switch to a new state
  if (pet.streak === 0) {
    const rnd = Math.random();
    let chosenSt = states.find((st) => rnd <= st.minProb);

    pet.state = chosenSt.name;
    pet.speed = chosenSt.speed;
    pet.minProb = chosenSt.minProb;
    pet.streak = chosenSt.streak;
  }

  pet.streak--;

  return {
    latitude: pos.latitude + randInt(-1, 1) * speedToDeg(pet.speed),
    longitude: pos.longitude + randInt(-1, 1) * speedToDeg(pet.speed),
  };
}

function randInt(min, max) {
  // min and max included
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function speedToDeg(speed) {
  const dist = speed * (cfg.LoRaWan.delay / 3600);
  return dist / 111; // Convert distance from kmh to lat/long degrees
}

module.exports = { getPeterPos };


novlog.js
const fs = require("fs");

const logsFolder = "./logs/";
const filePrefix = "novlog";

const LogType = {
  Info: "INFO",
  Error: "ERROR",
  Request: "REQUEST",
};

function log(data, logType) {
  const now = new Date();
  const logMsg = `[${getTimeAsString(now)}] [${logType}] - ${data}`;

  console.log(logMsg);

  // Write log data to a daily file in ./logs folder
  const dailyLogFile = `${filePrefix}${getDateAsString(now)}.log`;

  fs.appendFile(logsFolder + dailyLogFile, `${logMsg}\n`, (err) => {
    if (err) {
      throw err;
    }
  });
}

function info(msg) {
  log(msg, LogType.Info);
}

function error(msg) {
  log(msg, LogType.Error);
}

function request(req) {
  log(
    `[${req.ip}] ${req.method} ${req.originalUrl} with body ${JSON.stringify(
      req.body
    )}`,
    LogType.Request
  );
}

function getTimeAsString(time) {
  const cDay = zpad(time.getDate());
  const cMonth = zpad(time.getMonth() + 1);
  const cYear = zpad(time.getFullYear());

  const cHours = zpad(time.getHours());
  const cMinutes = zpad(time.getMinutes());
  const cSeconds = zpad(time.getSeconds());

  return `${cDay}/${cMonth}/${cYear} ${cHours}:${cMinutes}:${cSeconds}`;
}

function getDateAsString(date) {
  const cDay = zpad(date.getDate());
  const cMonth = zpad(date.getMonth() + 1);
  const cYear = date.getFullYear();
  return `${cDay}${cMonth}${cYear}`;
}

// Zero padding
function zpad(number) {
  return number < 10 ? `0${number}` : `${number}`;
}

module.exports = { error, info, request };


validator.js
/**
 * Username validation:
 * - Can only contains alphanumeric characters as well as "-" and "_"
 * - Must be between 3 and 30 characters long
 */
function isUsernameValid(username) {
  return username && username.match(/^[a-zA-Z0-9_-]{3,30}$/);
}

function isEmailValid(email) {
  return email && email.match(/[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+/);
}

function isPasswordValid(password) {
  return (
    password &&
    password.match(
      /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$ %^&*-]).{8,50}$/
    )
  );
}

// The eui must be a valid hexadecimal value of length 16
function isEUIValid(eui) {
  return eui && eui.match(/^[0-9a-fA-F]{16}$/);
}

function validateUserRegisterForm(username, email, password, callback) {
  let err;

  if (!isUsernameValid(username)) {
    err = "Username format invalid";
  } else if (!isEmailValid(email)) {
    err = "Email format invalid";
  } else if (!isPasswordValid(password)) {
    err = "Password format invalid";
  }

  callback(err);
}

function validateUserAuthenticateForm(username, password, callback) {
  let err;

  if (!isUsernameValid(username) && !isEmailValid(username)) {
    err = "Username/Email format invalid";
  } else if (!isPasswordValid(password)) {
    err = "Password format invalid";
  }

  callback(err);
}

module.exports = {
  validateUserRegisterForm,
  validateUserAuthenticateForm,
  isEUIValid,
};


index.js
// Node dependencies
const axios = require("axios");
const fs = require("fs");

const express = require("express");
const cors = require("cors");
const bodyParser = require("body-parser");
var session = require("express-session");

const bcrypt = require("bcrypt");

// Local dependencies
const validator = require("./validator.js");
const nv = require("./novlog.js");
const pt = require("./peter.js");

// Load configuration and data files
const config = require("./config.json");
const credentials = require("./credentials.json");
const db_users = require("./data/users.json");
const db_devices = require("./data/devices.json");

const app = express();
app.use(bodyParser.urlencoded({ extended: true }));

// Use the session middleware
app.use(
  cors({
    origin: `${config.webapp.host}:${config.webapp.port}`,
    credentials: true,
  }),
  express.json(), // automatically parse requests workload as JSON
  session({
    secret: config.secret,
    resave: false,
    saveUninitialized: true,
    cookie: { maxAge: 1800000 }, // 1800000 is equal to 30 minutes before a cookie expires
  })
);

// !insecure way : deactivating certificate checks
process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0;

// proper way to allow CampusIot -> download SSL certificate
//require("https").globalAgent.options.ca = require("ssl-root-cas/latest").create();

// Get JWT token asynchronously
// ! Since it takes a little bit of time to get the JWT, it is better to wait a little bit before issuing any API method (especially ones involving the LoRaWan server)
let JWT_TOKEN;
getJWT().then((data) => {
  JWT_TOKEN = data;
});

app.get("/", (req, res) => {
  nv.request(req);
  res.send("Hello World! Go to /api to use the api");
});

app.get("/api", (req, res) => {
  nv.request(req);
  let msg = "The API is up and running!<br/>Supported methods are:<br/>";
  msg += "POST /api/users/register<br/>";
  msg += "...<br/>";
  res.send(msg);
});

/**
 * @api [post] /api/register
 * description: "Creates a new user in the database matching the information given by the user"
 * parameters:
 *  - (query) username  {String} The username of the user
 *  - (query) email     {String} The email adress of the user
 *  - (query) password  {String} The password of the user
 * responses:
 *  200:
 *    description: The user has been succesfully registered
 *  400:
 *    description: The user could not be registered
 */
app.post("/api/register", (req, res) => {
  nv.request(req);
  // Get username, email and password from POST body
  const username = req.body.username;
  const email = req.body.email;
  const password = req.body.password;

  // Validate the form information entered by the user
  validator.validateUserRegisterForm(username, email, password, (err) => {
    // if the form information is invalid, send the error to the frontend
    if (err) {
      nv.info(`Register form error: ${err}`);
      res.status(400).json({
        status: "ERROR",
        error: err,
      });
      return;
    }

    if (db_users.find((user) => user.username === username)) {
      nv.info("Register form error: Username already taken");
      res.status(400).json({
        status: "ERROR",
        error: "Username already taken",
      });
      return;
    }

    if (db_users.find((user) => user.email === email)) {
      nv.info("Register form error: Email already in use");
      res.status(400).json({
        status: "ERROR",
        error: "Email already in use",
      });
      return;
    }

    bcrypt.genSalt(config.saltRounds, function (err, salt) {
      if (err) {
        nv.error(err);
        return;
      }

      bcrypt.hash(password, salt, function (err, hash) {
        if (err) {
          nv.error(err);
          res.status(400).json({
            status: "ERROR",
            error: "Unknown error during registration",
          });
          return;
        }

        // create the new user
        const niu = {
          username: username,
          email: email,
          password: hash,
          devices: [],
        };
        db_users.push(niu);

        writeJSONToFile(db_users, "./data/users.json");
      });
    });

    nv.info(`User ${username} successfully registered`);
    res.status(200).json({
      status: "OK",
    });
  });
});

app.post("/api/authenticate", (req, res) => {
  nv.request(req);
  const username = req.body.username;
  const password = req.body.password;

  // Validate the form information entered by the user
  validator.validateUserAuthenticateForm(username, password, (err) => {
    // if the form information is invalid, send the error to the frontend
    if (err) {
      nv.info(`Authentication form error: ${err}`);
      res.status(400).json({
        status: "ERROR",
        error: err,
      });
      return;
    }

    const user = db_users.find(
      (usr) => usr.username === username || usr.email === username
    );

    if (!user) {
      nv.info(`Unregistered user ${username} attempted to login`);
      res.status(400).json({
        status: "ERROR",
        error: "Invalid login",
      });
      return;
    }

    // Load hash from your password DB.
    bcrypt.compare(password, user.password, function (err, result) {
      if (result) {
        req.session.user = user.username;
        nv.info(`Successfull authentication for ${username}`);
        res.status(200).json({
          status: "OK",
        });
      } else {
        nv.error(err);
        res.status(400).json({
          status: "ERROR",
          error: "Invalid login",
        });
      }
    });
  });
});

app.post("/api/logout", (req, res) => {
  nv.request(req);

  if (!verifyAuthentication(req, res)) return;

  if (req.session) {
    const user = req.session.user;
    req.session.destroy((err) => {
      if (err) {
        nv.error(`Failed to delete session for ${user}`);
        res.status(400).json({
          status: "ERROR",
          error: "Failed to logout",
        });
      } else {
        nv.error(`Successfully deleted session for ${user}`);
        res.status(200).clearCookie("connect.sid").json({ status: "OK" });
      }
    });
  }
});

/**
 * How likely is it that an attacker making API calls with all possible EUIs
 * manage to register random device on the globe? We need to address this problem:
 * - Notify all users who have registered a device when a new user registers it
 * - Secure each device with a factory password to decrease the likeliness of random guessing
 * - Limit API usage to prevent a user from making thousands of calls per second
 * - Let it be as is (Since there are 16^16 possibles EUIs, it should take an average computer 500 years to try them all)
 */
app.post("/api/devices/:eui([0-9a-fA-F]{16})", (req, res) => {
  nv.request(req);

  if (!verifyAuthentication(req, res)) return;

  const user = db_users.find((usr) => usr.username === req.session.user);

  if (!user) {
    // TODO: I dunno what to do, this should never happen
    nv.error(
      "Database corruption: previously authenticated user is no longer a valid user"
    );
    res.status(456).json({
      status: "ERROR",
      error: "Database got corrupted, please contact an administrator",
    });
    return;
  }

  const eui = req.params.eui;

  // TODO: Not sure this check is relevant since Express already filters the route based on a regex
  if (!validator.isEUIValid(eui)) {
    res.status(400).json({
      status: "ERROR",
      error: "EUI format invalid",
    });
    return;
  }

  if (user.devices.find((devEUI) => devEUI === eui)) {
    nv.info(`Device ${eui} already registered by ${user.username}`);
    res.status(400).json({
      status: "ERROR",
      error: "Device already registered",
    });
    return;
  }

  axios
    .get(`${config.LoRaWan.endpoint.host}/devices/${eui}`, {
      headers: { "Grpc-Metadata-Authorization": `Bearer ${JWT_TOKEN}` },
    })
    .then(() => {
      user.devices.push(eui);
      writeJSONToFile(db_users, "./data/users.json");

      let device = db_devices.find((dev) => dev.eui === eui);

      if (!device) {
        db_devices.push({
          eui: eui,
          owners: [user.username],
          positions: [],
        });
      } else {
        device.owners.push(user.username);
      }
      writeJSONToFile(db_devices, "./data/devices.json");
      nv.info(`Device ${eui} successfully registered by ${user.username}`);
      res.status(200).json({
        status: "OK",
      });
    })
    .catch((err) => {
      nv.error(err);
      res.status(400).json({
        status: "ERROR",
        error: "Device could not be registered",
      });
    });

  // retrieve user from cookie
  // check if device is a valid tracker
  // check if device has not already been registered by a user
  // eventually protect devices with default password that can be changed to prevent a user to re-register a device that is not their own
  // if device and user are legitimate, return success
});

app.delete("/api/devices/:eui([0-9a-fA-F]{16})", (req, res) => {
  nv.request(req);

  if (!verifyAuthentication(req, res)) return;

  const user = db_users.find((usr) => usr.username === req.session.user);

  if (!user) {
    // TODO: I dunno what to do, this should never happen
    nv.error(
      "Database corruption: previously authenticated user is no longer a valid user"
    );
    res.status(456).json({
      status: "ERROR",
      error: "Database got corrupted, please contact an administrator",
    });
    return;
  }

  const eui = req.params.eui;

  // TODO: Not sure this check is relevant since Express already filters the route based on a regex
  if (!validator.isEUIValid(eui)) {
    res.status(400).json({
      status: "ERROR",
      error: "EUI format invalid",
    });
    return;
  }

  if (!user.devices.find((devEUI) => devEUI === eui)) {
    nv.info(`Device ${eui} not registered by ${user.username}`);
    res.status(400).json({
      status: "ERROR",
      error: "Device has not been registered",
    });
    return;
  }

  user.devices = user.devices.filter((devEUI) => devEUI !== eui);
  writeJSONToFile(db_users, "./data/users.json");

  let device = db_devices.find((dev) => dev.eui === eui);

  if (!device) {
    // TODO: I dunno what to do, this should never happen
    nv.error(
      "Database corruption: previously registered device is no longer a valid device"
    );
    res.status(456).json({
      status: "ERROR",
      error: "Database got corrupted, please contact an administrator",
    });
    return;
  } else {
    device.owners = device.owners.filter((usr) => usr !== user.username);
  }
  writeJSONToFile(db_devices, "./data/devices.json");
  nv.info(`Device ${eui} successfully deleted for ${user.username}`);
  res.status(200).json({
    status: "OK",
  });
});

app.get("/api/devices", (req, res) => {
  nv.request(req);

  if (!verifyAuthentication(req, res)) return;

  const user = db_users.find((usr) => usr.username === req.session.user);

  if (!user) {
    // TODO: I dunno what to do, this should never happen
    nv.error(
      "Database corruption: previously authenticated user is no longer a valid user"
    );
    res.status(456).json({
      status: "ERROR",
      error: "Database got corrupted, please contact an administrator",
    });
    return;
  }

  nv.info(`Successfully fetched devices for ${user.username}`);
  res.status(200).json({
    devices: user.devices,
  });
});

app.get("/api/devices/:eui([0-9a-fA-F]{16})", (req, res) => {
  nv.request(req);

  if (!verifyAuthentication(req, res)) return;

  const user = db_users.find((usr) => usr.username === req.session.user);

  if (!user) {
    // TODO: I dunno what to do, this should never happen
    nv.error(
      "Database corruption: previously authenticated user is no longer a valid user"
    );
    res.status(456).json({
      status: "ERROR",
      error: "Database got corrupted, please contact an administrator",
    });
    return;
  }

  const eui = req.params.eui;

  // TODO: Not sure this check is relevant since Express already filters the route based on a regex
  if (!validator.isEUIValid(eui)) {
    res.status(400).json({
      status: "ERROR",
      error: "EUI format invalid",
    });
    return;
  }

  if (!user.devices.find((devEUI) => devEUI === eui)) {
    nv.info(`Device ${eui} not registered by ${user.username}`);
    res.status(400).json({
      status: "ERROR",
      error: "Device has not been registered",
    });
    return;
  }

  let device = db_devices.find((dev) => dev.eui === eui);

  if (!device) {
    // TODO: I dunno what to do, this should never happen
    nv.error(
      "Database corruption: previously registered device is no longer a valid device"
    );
    res.status(456).json({
      status: "ERROR",
      error: "Database got corrupted, please contact an administrator",
    });
    return;
  }

  // Get current time in Unix Timestamp format
  const now = new Date().getTime();
  const latestStoredPos = device.positions[0];

  // If the latest stored position is too recent, send this position to the frontend client
  if (
    latestStoredPos &&
    now - latestStoredPos.timestamp < config.LoRaWan.delay * 1000
  ) {
    nv.info(
      `Successfully fetched last stored position for device ${device.eui} from server cache`
    );
    res.status(200).json(latestStoredPos);
    return;
  }

  // The latest position is too old, make an API call to the LoRaWan endpoint to obtain the latest position
  /*  const latestPos = {
    latitude: 45.20415,
    longitude: 5.6933013,
    timestamp: now,
  }; // TODO: Replace with a call to CampusIoT to get POS */
  const latestPos = pt.getPeterPos(); // call to Peter module to get fake card data

  device.positions.unshift(latestPos);
  if (device.positions.length > config.deviceMaxPos) {
    device.positions.pop();
  }

  writeJSONToFile(db_devices, "./data/devices.json");
  nv.info(
    `Successfully retrieved latest known position for device ${device.eui} from LoRaWan endpoint`
  );
  res.status(200).json(latestPos);

  /* axios
    .get(`${config.LoRaWan.endpoint.host}/devices/${eui}`, {
      headers: { "Grpc-Metadata-Authorization": `Bearer ${JWT_TOKEN}` },
    })
    .then((res2) => {
      console.log(res2);
      res.send(res2.data.device.description);
    })
    .catch((err) => {
      console.log(err);
      res.send("An error occured");
    }); */
});

app.use(function (req, res) {
  nv.request(req);
  nv.info(`Page not found: ${req.originalUrl}`);
  res.status(404).json({
    status: "ERROR",
    error: "Page not found",
  });
});

app.listen(config.port, () => {
  nv.info(`psp-backend listening at http://localhost:${config.port}`);
});

function getJWT() {
  // POST request to LoRaWan endpoint with group credentials to obtain a valid JWT
  return axios
    .post(config.LoRaWan.endpoint.host + "/internal/login", credentials)
    .then((res) => {
      nv.info(`JWT token successfully retrieved: ${res.data.jwt}`);
      return res.data.jwt;
    })
    .catch((error) => {
      nv.error(error);
      return null;
    });
}

function writeJSONToFile(obj, path) {
  // Convert JSON object to a string
  const data = JSON.stringify(obj);

  // Attempt to write file to disk
  fs.writeFile(path, data, "utf8", (err) => {
    if (err) {
      nv.error(`Error writing data to ${path}: ${err}`);
    } else {
      nv.info(`Successfully updated data in ${path}!`);
    }
  });
}

function verifyAuthentication(req, res) {
  if (!req.session.user) {
    nv.info("Authentication required");
    res.status(401).json({
      status: "ERROR",
      error: "Authentication required",
    });
    return false;
  }

  return true;
}
